### PR TITLE
Add neon.primary_is_running GUC.

### PR DESCRIPTION
**Pull request by lubennikovaav**
We set it for neon replica, if primary is running

Corresponding cloud PR is https://github.com/neondatabase/cloud/pull/10183

### CODE REVIEW

**Review by knizhnik — APPROVED** [review]

**Review by arpad-m — APPROVED** [review]

**Comment by hlinnaka** [comment]
For the sake of the archives, please explain the problem in the commit message too. And perhaps open a github issue to track it. IIUC the problem is that:

- No primary is running on a branch
- You start a new read-only replica on the branch
- The read-only replica doesn't allow connections, because it doesn't have a valid standby snapshot yet. It doesn't know which transactions that have not committed yet might still be running in the primary and commit later. There is no primary, so the answer is "none", but Postgres doesn't know that.

A hot standby will wait for RUNNING_XACTS or shutdown checkpoint records until it can allow connections. If the primary is not running, they will never come.

This PR adds a GUC to tell the standby that there is no primary running. It doesn't actually do anything with it. Will that be a followup PR?

Even with this flag, this can still happen:

- There _is_ a primary running
- You start a read-only replica
- Primary crashes / is killed before it writes a RUNNING_XACTS record.

So just checking that a primary is running when the standby starts isn't necessarily enough.

A GUC doesn't feel like a very good mechanism to deliver this information. Whether a primary is running or not is very ephemeral. Perhaps the neon extension should ask the control plane for that directly with an API call?

File: libs/compute_api/src/spec.rs [hunk 1]
```diff
@@ -79,6 +79,12 @@ pub struct ComputeSpec {
     // Stripe size for pageserver sharding, in pages
     #[serde(default)]
     pub shard_stripe_size: Option<usize>,
+
+    // When we are starting a new replica in hot standby mode,
+    // we need to know if the primary is running.
+    // This is used to determine if we should try to pull SLRU from primary
```
Comment (hlinnaka) [inline libs/compute_api/src/spec.rs:85]: This has nothing to do with SLRUs

**Comment by knizhnik** [comment]
> For the sake of the archives, please explain the problem in the commit message too. And perhaps open a github issue to track it. IIUC the problem is that:
> 
> * No primary is running on a branch
> * You start a new read-only replica on the branch
> * The read-only replica doesn't allow connections, because it doesn't have a valid standby snapshot yet. It doesn't know which transactions that have not committed yet might still be running in the primary and commit later. There is no primary, so the answer is "none", but Postgres doesn't know that.
> 
> A hot standby will wait for RUNNING_XACTS or shutdown checkpoint records until it can allow connections. If the primary is not running, they will never come.
> 
> This PR adds a GUC to tell the standby that there is no primary running. It doesn't actually do anything with it. Will that be a followup PR?


Hot-standby replica  is waiting for primary only if was not started after normal shutdown (wasShutdown true in xlogrecovery.c). This behaviour was changed only by my PR #6357. And I am also going to handle added GUC in this PR.

> 
> Even with this flag, this can still happen:
> 
> * There _is_ a primary running
> * You start a read-only replica
> * Primary crashes / is killed before it writes a RUNNING_XACTS record.

Primary has to be restarted in any case. 
We do not support replacement of primary with replica.
So Neon should detect crash of primary node and restart it.
Replica will reconnect to new primary and receive  RUNNING_XACTS record.

Actually replica has no other choice rather than wait for primary once it starts recovery - receiving WAL from primary.
AS a result it can receive tuples which can contain XIDs of running transactions. This transactions can be committed before primary crash. So replica may not set XMIN_INVSLID hint but for such tuples.

> 
> So just checking that a primary is running when the standby starts isn't necessarily enough.

I do not know better solution.

> 
> A GUC doesn't feel like a very good mechanism to deliver this information. Whether a primary is running or not is very ephemeral. Perhaps the neon extension should ask the control plane for that directly with an API call?

The code making decision whether to set `wasInterrupted` to not was not in extension, but in Postgres core (xlogrecovery.c)
Certainly it is possible to add one more hook here. But we have to change Postgres sources in any case.

I do not understand why sending request to control plane is better than sending this information through compute spec and then GUC. Also extra round-trip doesn't speedup startup.

**Comment by lubennikovaav** [comment]
> >Even with this flag, this can still happen:
> > There is a primary running
> You start a read-only replica
> Primary crashes / is killed before it writes a RUNNING_XACTS record.
> Primary has to be restarted in any case.

> We do not support replacement of primary with replica.
> So Neon should detect crash of primary node and restart it.
> Replica will reconnect to new primary and receive RUNNING_XACTS record.
> 

Hmm, looks like we can get into a deadlock here. Imagine this situation:
- primary is running
- replica is starting and waiting for running_xacts from primary
- primary is crashing
- StartCompute(replica) operation is still in progress, and control-plane cannot run concurrent operation of StartCompute(primary), so it will wait in a queue.
- Eventually, this situation will resolve, when replica startCompute deadline exceeds.

Are we OK with this?
I'd say yes, because crashing primary is not normal and should be fixed asap anyway.
@hlinnaka, @knizhnik, WDYT?

File: libs/compute_api/src/spec.rs (hunk 1 above)
Comment (lubennikovaav) [reply libs/compute_api/src/spec.rs]: thank you for pointing out!
I confused running_xacts record with xacts SLURU

**Comment by hlinnaka** [comment]
> > > Even with this flag, this can still happen:
> > > There is a primary running
> > > You start a read-only replica
> > > Primary crashes / is killed before it writes a RUNNING_XACTS record.
> > > Primary has to be restarted in any case.
> 
> > We do not support replacement of primary with replica.
> > So Neon should detect crash of primary node and restart it.
> > Replica will reconnect to new primary and receive RUNNING_XACTS record.
> 
> Hmm, looks like we can get into a deadlock here. Imagine this situation:
> 
>     * primary is running
> 
>     * replica is starting and waiting for running_xacts from primary
> 
>     * primary is crashing
> 
>     * StartCompute(replica) operation is still in progress, and control-plane cannot run concurrent operation of StartCompute(primary), so it will wait in a queue.
> 
>     * Eventually, this situation will resolve, when replica startCompute deadline exceeds.
> 
> 
> Are we OK with this? I'd say yes, because crashing primary is not normal and should be fixed asap anyway. @hlinnaka, @knizhnik, WDYT?

Until https://github.com/neondatabase/neon/pull/6712 is merged, we don't actually write the shutdown checkpoint record even on a clean shutdown, so you'll get the above with a clean shutdown too.

**Comment by hlinnaka** [comment]
> > This PR adds a GUC to tell the standby that there is no primary running. It doesn't actually do anything with it. Will that be a followup PR?
> 
> Hot-standby replica is waiting for primary only if was not started after normal shutdown (wasShutdown true in xlogrecovery.c). This behaviour was changed only by my PR #6357. And I am also going to handle added GUC in this PR.

I see. Can you include those changes here, please, so that this can be reviewed, tested and committed as one unit, please?

Needs tests.

**Comment by knizhnik** [comment]
> So we currently have this bug:
> 
> 1. Primary is running
> 2. Start a transaction in primary. Insert a row.
> 3. Start a replica. Because we always set 'wasShutdown==true' in xlogrecovery.c, the replica considers all in-progress transactions as aborted.
> 4. On replica. `BEGIN REPEATABLE READ;` and run a query on replica. The new row inserted in primary is not visible.
> 5. Commit the transaction on primary
> 6. On replica: run another query in the same transaction. Incorrectly, the new row is now visible.
> 
> Please create a python test case for that. This PR should then fix it.

There is such test  test_runner/regress/test_replication_start.py in PR #6357:
https://github.com/neondatabase/neon/pull/6357/files#diff-7edc60d2be535ca5bbd3cce1b149594fd1994e755bd3b1f362a7e010a5cd3b16

**Comment by knizhnik** [comment]
> > > This PR adds a GUC to tell the standby that there is no primary running. It doesn't actually do anything with it. Will that be a followup PR?
> > 
> > 
> > Hot-standby replica is waiting for primary only if was not started after normal shutdown (wasShutdown true in xlogrecovery.c). This behaviour was changed only by my PR #6357. And I am also going to handle added GUC in this PR.
> 
> I see. Can you include those changes here, please, so that this can be reviewed, tested and committed as one unit, please?
> 
> Needs tests.

I can shuffle changes between PR as you prefer.
But actually this problem and its fix has no relation to this PR.
So, what we have no: replica is started with `wasShutdown=true` doesn't receive information about running xacts, construct incorrect snapshot and as a result see incorrect data.
The problem was fixed in #6357 - now hot-standby replica waits for WAL from primary.
It cause problem with e2e tests (which spawns read-only replica on branch without spawning master). But certainly it is not only related to e2e. So Star has proposed this mechanism with informing replica by control place whether primary is running (wasShutdown=false) or not (wasShutdown=true).

So #6357 solves the problem with spawning replica with alive master. And this PR (in conjunction with correspondent PR for control plane) solves the problem with spawning replica without alive master.
Please notice that
1. To make this POR work we need first merge it's control plane part
2. Regression tests are not using control plane. So to test it we do not need this PR. It is possible to propagate this flag through `neon_local`. But I am not sure that it makes sense.

**Comment by hlinnaka** [comment]
> I can shuffle changes between PR as you prefer.

Thank you. The general rule should be: one bug, one PR. PR includes all code changes required to fix the bug, and a test to reproduce it, but nothing else.

> But actually this problem and its fix has no relation to this PR. So, what we have no: replica is started with `wasShutdown=true` doesn't receive information about running xacts, construct incorrect snapshot and as a result see incorrect data.

The point of this PR is to fix that bug, right?

> The problem was fixed in #6357 - now hot-standby replica waits for WAL from primary.

According to the description of #6357, it is about propagating apply_lsn from safekeepers to pageservers, to postpone GC. That seems completely unrelated to this.

> It cause problem with e2e tests (which spawns read-only replica on branch without spawning master). But certainly it is not only related to e2e. So Star has proposed this mechanism with informing replica by control place whether primary is running (wasShutdown=false) or not (wasShutdown=true).
> 
> So #6357 solves the problem with spawning replica with alive master. And this PR (in conjunction with correspondent PR for control plane) solves the problem with spawning replica without alive master. Please notice that
> 
>     1. To make this POR work we need first merge it's control plane part

Gotcha.

>     2. Regression tests are not using control plane. So to test it we do not need this PR. It is possible to propagate this flag through `neon_local`. But I am not sure that it makes sense.

Yes, it does. We need regression tests for these things. If we go with a GUC, it's easy to set the GUC in a test to simulate how the control plane would set it.

To summarize, for this PR:

- Let's agree that the point of this PR is to fix the bug that a replica constructs incorrect snapshots, which leads to incorrect query results, if the primary is running.
- Please include all the code changes needed to fix _that_ bug, but nothing else.
- Please include a regression test case that demonstrates the bug. Without the code changes from PR, the test fails, and with the changes, it passes.

**Comment by knizhnik** [comment]
> > I can shuffle changes between PR as you prefer.
> 
> Thank you. The general rule should be: one bug, one PR. PR includes all code changes required to fix the bug, and a test to reproduce it, but nothing else.
> 
> > But actually this problem and its fix has no relation to this PR. So, what we have no: replica is started with `wasShutdown=true` doesn't receive information about running xacts, construct incorrect snapshot and as a result see incorrect data.
> 
> The point of this PR is to fix that bug, right?
> 
> > The problem was fixed in #6357 - now hot-standby replica waits for WAL from primary.
> 
> According to the description of #6357, it is about propagating apply_lsn from safekeepers to pageservers, to postpone GC. That seems completely unrelated to this.
> 
> > It cause problem with e2e tests (which spawns read-only replica on branch without spawning master). But certainly it is not only related to e2e. So Star has proposed this mechanism with informing replica by control place whether primary is running (wasShutdown=false) or not (wasShutdown=true).
> > So #6357 solves the problem with spawning replica with alive master. And this PR (in conjunction with correspondent PR for control plane) solves the problem with spawning replica without alive master. Please notice that
> > 
> > 1. To make this POR work we need first merge it's control plane part
> 
> Gotcha.
> 
> > 2. Regression tests are not using control plane. So to test it we do not need this PR. It is possible to propagate this flag through `neon_local`. But I am not sure that it makes sense.
> 
> Yes, it does. We need regression tests for these things. If we go with a GUC, it's easy to set the GUC in a test to simulate how the control plane would set it.
> 
> To summarize, for this PR:
> 
> * Let's agree that the point of this PR is to fix the bug that a replica constructs incorrect snapshots, which leads to incorrect query results, if the primary is running.
> * Please include all the code changes needed to fix _that_ bug, but nothing else.
> * Please include a regression test case that demonstrates the bug. Without the code changes from PR, the test fails, and with the changes, it passes.

Done

File: control_plane/src/endpoint.rs [hunk 2]
```diff
@@ -290,6 +290,7 @@ impl Endpoint {
         conf.append("wal_log_hints", "off");
         conf.append("max_replication_slots", "10");
         conf.append("hot_standby", "on");
+        conf.append("hot_standby_feedback", "on");
```
Comment (hlinnaka) [inline control_plane/src/endpoint.rs:293]: unrelated change

File: libs/postgres_ffi/src/xlog_utils.rs [hunk 3]
```diff
@@ -119,11 +119,6 @@ pub fn generate_pg_control(
     // Generate new pg_control needed for bootstrap
     checkpoint.redo = normalize_lsn(lsn, WAL_SEGMENT_SIZE).0;
 
-    //reset some fields we don't want to preserve
-    //TODO Check this.
-    //We may need to determine the value from twophase data.
-    checkpoint.oldestActiveXid = 0;
-
```
Comment (hlinnaka) [inline libs/postgres_ffi/src/xlog_utils.rs:126]: Do we store correct and track the oldestActiveXid value correctly in the pageserver now? I think not. Looking at the code in walingest.rs, we ignore all other fields from checkpoint records except oldestXid.

File: pgxn/neon/neon.c [hunk 4]
```diff
@@ -181,6 +183,15 @@ _PG_init(void)
 
 	pg_init_extension_server();
 
+	DefineCustomBoolVariable(
+		"neon.primary_is_running",
+		"For replica it is true, if primary is running, false otherwise",
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pgxn/neon/neon.c:188]: 'true' would be the safer default

File: test_runner/regress/test_replication_start.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+                with env.endpoints.new_replica_start(
+                    origin=primary, endpoint_id="secondary"
+                ) as secondary:
+                    wait_replica_caughtup(primary, secondary)
+                    with secondary.connect() as s_con:
+                        with s_con.cursor() as s_cur:
+                            s_cur.execute("select * from pg_class")
+                            p_cur.execute("commit")
+                            wait_replica_caughtup(primary, secondary)
+                            s_cur.execute("select * from t where pk = 1")
```
Comment (hlinnaka) [inline test_runner/regress/test_replication_start.py:26]: You're not checking the results of these queries. I presume this test will pass with or without the code changes, which is not good. Please make sure the test fails without the fix.

Comment (knizhnik) [reply test_runner/regress/test_replication_start.py:26]: Without this fix table 't' is not failed at replica and test failed with error.

File: libs/postgres_ffi/src/xlog_utils.rs (hunk 3 above)
Comment (knizhnik) [reply libs/postgres_ffi/src/xlog_utils.rs:126]: Add tracking also of `oldestActriveXid`

File: test_runner/regress/test_replication_start.py (hunk 5 above)
Comment (knizhnik) [reply test_runner/regress/test_replication_start.py:26]: But added check in any case...

File: compute_tools/src/config.rs [hunk 6]
```diff
@@ -79,6 +79,14 @@ pub fn write_postgres_conf(
...
             // hot_standby is 'on' by default, but let's be explicit
             writeln!(file, "hot_standby=on")?;
+
+            // Inform the replica about the primary state
+            // Default is 'false'
+            writeln!(
+                file,
+                "neon.primary_is_running={}",
+                spec.primary_is_running.unwrap_or(false)
+            )?;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply compute_tools/src/config.rs]: Are you sure that it is safe default?
If we assume that primary is running but it is not, then startup of hot-standby replica hangs forever.
That is what happening now with e2e tests.
May be we should assume that if information about primary is not available, then it is not active?
In this case this PR should pass e2e tests even without changes in control plane?

File: test_runner/regress/test_replication_start.py (hunk 5 above)
Comment (hlinnaka) [reply test_runner/regress/test_replication_start.py:26]: Please verify that this test fails without the code changes. I'm still not 100% convinced.

By the time you run the `select * from t where pk = 1` query, the standby has received the COMMIT record, and sees the transaction as visible, with or without this bug fix. The problem scenario we're fixing is the other way round. If you start a transaction and acquire a snapshot in the replica before the commit, and then use the snapshot after the COMMIT record has been received, the snapshot should still see the transaction as in-progress.

File: libs/postgres_ffi/src/xlog_utils.rs (hunk 3 above)
Comment (hlinnaka) [reply libs/postgres_ffi/src/xlog_utils.rs:126]: Ok, please add a test that fails when you remove the tracking.

File: test_runner/regress/test_replication_start.py (hunk 5 above)
Comment (knizhnik) [reply test_runner/regress/test_replication_start.py:26]: I confirm that this test failed without this change.
So what's happening: 
- replica receives INSERT WAL record in pg_class for created table 't'. Transaction is not committed yet
- we run query `select * from pg_class` at replica which set XMIN_INVALID hint bit for this tuple
- replica receives COMMIT record. But tuple already has XMIN_INVALID hint bit and so is not visible by MVCC

File: compute_tools/src/config.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hlinnaka) [reply compute_tools/src/config.rs]: By "safe", I meant "you won't get incorrect query results". You're right that hanging forever is not nice either, and may even be worse. Still, I'd be more comfortable waiting than possibly returning wrong results.

File: libs/postgres_ffi/src/xlog_utils.rs (hunk 3 above)
Comment (knizhnik) [reply libs/postgres_ffi/src/xlog_utils.rs:126]: If I do not track `oldestActiveXid`, then assertion failure will happen in xlog.c:
```
			if (wasShutdown)
				oldestActiveXID = PrescanPreparedTransactions(&xids, &nxids);
			else
				oldestActiveXID = checkPoint.oldestActiveXid;
			Assert(TransactionIdIsValid(oldestActiveXID));
```
if I assign FirstNormalTransactionId to checkPoint.oldestActiveXid, then it is most conservative estimation of oldest xid and so it can not cause any errors. Well, it is not true in case of xid  wraparound (xid > 2^31).  But it is hard to cause wraparound in the test.

File: compute_tools/src/config.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply compute_tools/src/config.rs]: Well, I agree with you.
I just want to postpone this change until https://github.com/neondatabase/cloud/pull/10183 is merged.
It gives a chance to e2e tests to pass.

[the remark this answers is not in the document]
Comment (hlinnaka) [reply compute_tools/src/config.rs]: Fair. Alternatively, you could merge https://github.com/neondatabase/cloud/pull/10183 first. That one passes the tests without this PR. I'd prefer to commit this in one piece, otherwise it's hard to track what issues remain.

Do we need any special considerations in how this is rolled out to production? We need to make sure the console change are deployed first?

[the remark this answers is not in the document]
Comment (knizhnik) [reply compute_tools/src/config.rs]: So https://github.com/neondatabase/cloud/pull/10183 is now merged.
What we should do then?
It seems to me that it is still "safer" to keep default value of primary_is_running=false, because it corresponds to the current status quo: when we start any compute mode, we assume that it happens after normal shutdown.
It also simplify rollup: with default=false we can deploy control plane and compute in any order:
Control plain includes "primary_is_running" is spec, but it will not be transformed into GUC until compute is deployed. And visa versa - if we deploy first compute, then with default value `primary_is_running=false`
it preserves the current behaviour.

We can always change default in future.

File: pageserver/src/walingest.rs [hunk 7]
```diff
@@ -334,6 +334,20 @@ impl WalIngest {
...
+                    );
+                    if (self
+                        .checkpoint
+                        .oldestActiveXid
+                        .wrapping_sub(xlog_checkpoint.oldestActiveXid)
+                        as i32)
+                        < 0
+                    {
+                        self.checkpoint.oldestActiveXid = xlog_checkpoint.oldestActiveXid;
+                    }
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/walingest.rs]: Done

**Comment by skyzh** [comment]
ref https://github.com/neondatabase/neon/issues/7204
